### PR TITLE
shallow cloning + GitHub Action

### PR DESCRIPTION
**Pull request by joeleonjr**
### Description:
This proposed GitHub Action updates the existing action by implementing a few new features:

1. The action defaults to using git shallow cloning. In most CI/CD use cases, users only want to scan the code diff in a push or PR and not the entire repo. By performing a shallow clone, we only clone down the relevant commits (and files). From there we invoke TruffleHog to scan only code diffs, thus creating a more performant action.
2. The action includes git cloning (using the default `actions/checkout@v3` action). Why? Two reasons: (A) user experience: users can now invoke this action in their GH workflow yaml files without any additional steps
```
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
   - name: TruffleHog OSS
        id: trufflehog
        uses: trufflesecurity/trufflehog@v2
        with:
           extra_args: --only-verified
```
(B) Git shallow cloning requires a few computations (ex: count of commits in the push or PR). This is simpler for the user; otherwise, they'd have to follow a *slightly* convoluted set-up process.
 3. The current action does not provide a sufficiently robust process to scan pushes to the default branch. In some workflows, engineers might push multiple commits inside one push to `main`. Given the current action's tutorial, I couldn't easily scan a push to `main` containing multiple commits. (Maybe just my user error though):

``` 
...
with:
    path: ./
    base: ${{ github.event.repository.default_branch }}
    head: HEAD
```
It appears that in order to scan a push with multiple commits, the user would need to provide the commit hash prior to the new commits into the  `base` argument. There doesn't seem to be a straightforward way to do that in the current action.

The updated action will do this by default.

4. The action allows users to scan their entire repo by passing in the `full_repo` flag (conducting a full git clone). This is for users that want to scan more than just the code diff in their push or PR.
5. The action removes references to `path`, `base` and `head`, which seem to confuse some users.

Note: To prevent backwards compatibility issues, I propose we tag the original action `v1` (it's currently `main`). Then we tag the new action `v2`. Users will still be able to use `v1` with an always-updated version of TH OSS, since we invoke the `latest` docker image. 

Comments and discussion are welcomed!! Also, I did my best to test this out, but it would be helpful for others to run this action inside larger repositories (alongside the existing action).

### Checklist:
* [x] Tests passing (`make test-community`)?
* [x] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

**Comment by rosecodym** [comment]
(1) and (3) seem like great improvements. I find myself skeptical of (2) and (5) - personally, I prefer CI actions that are explicit, granular, and as un-magical as possible. Batteries-included actions are great when they work, but as soon as they don't, I'm left with a paucity of options to debug.

But! I claim no authority for this opinion, and you interact with other developers who use this more than I do, so I'll defer to the judgment of the crowd.

**Comment by joeleonjr** [comment]
> (1) and (3) seem like great improvements. I find myself skeptical of (2) and (5) - personally, I prefer CI actions that are explicit, granular, and as un-magical as possible. Batteries-included actions are great when they work, but as soon as they don't, I'm left with a paucity of options to debug.
> 
> But! I claim no authority for this opinion, and you interact with other developers who use this more than I do, so I'll defer to the judgment of the crowd.

I think that's a totally fair response. The two main set-up difficulties that I'm trying to abstract for users is: (1) conducting a git shallow clone, and (2) counting commits in pushes, so that the entire push can be easily scanned. 

If the better option is to write up two new documentation sections and add a new flag (or two) to the existing GH Action structure, I'm happy to do that.

**Comment by codevbus** [comment]
> 2. The action includes git cloning (using the default actions/checkout@v3 action). Why? Two reasons: (A) user experience: users can now invoke this action in their GH workflow yaml files without any additional steps

This is the only one I have issue with. To echo @rosecodym; this is abstracting away the modularity/granularity of a composable CI/CD pipeline. Since we can generally assume most users will consume this as part of a larger pipeline, they're going to end up checking out their code twice.

That being said, it's also clear that's what's needed to abstract away the cloning options needed to make this work. I'm not sure there's an alternative and this may just need to be a compromise.

**Comment by zricethezav** [comment]
> Note: To prevent backwards compatibility issues, I propose we tag the original action v1 (it's currently main). Then we tag the new action v2. Users will still be able to use v1 with an always-updated version of TH OSS, since we invoke the latest docker image.

I think this is as good plan

**Comment by joeleonjr** [comment]
Just pushed a major update to the PR. A few changes:

1. This PR no longer embeds git cloning and instead provides users with instructions for shallow cloning in the README. 
2. It addresses the issue of users not scanning during pushes. If values inputted by a user for the `base` and `head` fields are equal, it will throw an error. The error message links to our README and explains that TruffleHog wouldn't have scanned anything since the `git diff` was null.
3. This PR is fully backward compatible, so no versioning needed. All prior flags still exist and work as expected.
4. Users can now implement the action with no flags. The default is to scan all code changes any time it runs (in `main`, a `feature-branch`, or any PR). If the action is invoked via `workflow_dispatch` or `schedule`, it will default to scanning the entire git repository history.

**Review by zricethezav — APPROVED** [review]
Just left a non-blocking comment for discussion. I think this PR is good to go. Huge huge huge improvement, thanks Joe!

File: action.yml
Comment (joeleonjr) [reply action.yml:51]: Definitely a valid use case. I'm open to adjusting this. I think we should ship this asap since some folks aren't scanning commits atm and think they are. But I'm down to think through what the behavior should look like.

Comment (joeleonjr) [reply action.yml:51]: Separately, I'm reviewing this code before merging in and I'm looking at this line `if [ $base_commit == $head_commit ] ; then`. Do you think there are folks that leave `base` equal to `''` and `head` equal to `''`, so that they can trigger a full git repository history scan? @zricethezav

**Comment by aldy505** [comment]
Hi @joeleonjr, it turns out this commit broke my workflow. I'm using self-hosted GitHub Actions runner for my organization, and I'm guarantee CI environment isolation by running the CI via Docker. This PR removes the Docker image and it fails on everyone who uses Docker [container base](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container) on their GitHub Actions.

Here's my GitHub Actions output:
image

I'm not saying we should revert this PR, as from what I read this is a good direction going forward. What we can do is providing a documentation on the README that points out that the current quickstart for GitHub Actions won't work if you're using a container base.

A small reproduction would be this (you can run this on both self-hosted runner and SaaS -- or GitHub-hosted runners):

```yaml
name: PR

on:
  pull_request:

jobs:
  ci:
    name: CI
    container: golang:bookworm # This is the part that defines the container base image
    runs-on: "ubuntu-latest"
    steps:
      - name: Checkout code
        uses: actions/checkout@v4
        with:
          fetch-depth: 0

      - name: TruffleHog OSS
        uses: trufflesecurity/trufflehog@main
        with:
          path: ./
          base: ${{ github.event.repository.default_branch }}
          head: HEAD
          extra_args: --debug --only-verified
```

**Comment by joeleonjr** [comment]
> Hi @joeleonjr, it turns out this commit broke my workflow. I'm using self-hosted GitHub Actions runner for my organization, and I'm guarantee CI environment isolation by running the CI via Docker. This PR removes the Docker image and it fails on everyone who uses Docker [container base](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container) on their GitHub Actions.
> 
> Here's my GitHub Actions output: image
> 
> I'm not saying we should revert this PR, as from what I read this is a good direction going forward. What we can do is providing a documentation on the README that points out that the current quickstart for GitHub Actions won't work if you're using a container base.
> 
> A small reproduction would be this (you can run this on both self-hosted runner and SaaS -- or GitHub-hosted runners):
> 
> ```yaml
> name: PR
> 
> on:
>   pull_request:
> 
> jobs:
>   ci:
>     name: CI
>     container: golang:bookworm # This is the part that defines the container base image
>     runs-on: "ubuntu-latest"
>     steps:
>       - name: Checkout code
>         uses: actions/checkout@v4
>         with:
>           fetch-depth: 0
> 
>       - name: TruffleHog OSS
>         uses: trufflesecurity/trufflehog@main
>         with:
>           path: ./
>           base: ${{ github.event.repository.default_branch }}
>           head: HEAD
>           extra_args: --debug --only-verified
> ```

Thanks for bringing this to our attention. I think I have a solution that would require an update to the TruffleHog action.yml file. I'm proposing a check to determine whether the action is already containerized. If so, then it will install golang, clone/build TruffleHog and then run TruffleHog. If not, it will just use the dockerized version. 

Do you mind @aldy505 letting me know if this works for your use case?

```
jobs:
  ci1:
    name: CI
    container: golang:bookworm # This is the part that defines the container base image
    runs-on: "ubuntu-latest"
    steps:
      - name: Checkout code
        uses: actions/checkout@v4
        with:
          fetch-depth: 0

      - name: Install Golang (Running in Container)
        if: ${{ job.container != '' }}
        uses: actions/setup-go@v4
        with:
          go-version: '1.21'

      - name: Install/Compile TruffleHog (Running in Container)
        if: ${{ job.container != '' }}
        run: |
          git clone https://github.com/trufflesecurity/trufflehog.git
          cd trufflehog; go install; cd -
      - name: Run TruffleHog (Running in Container)
        if: ${{ job.container != '' }}
        run: |
          trufflehog git file://./ \
          --since-commit ${{ github.event.repository.default_branch }} \
          --branch 'HEAD' \
          --fail \
          --no-update \
          --github-actions \
          --debug --trace \

      - name: Run TruffleHog (with Docker)
        if: ${{ job.container == '' }}
        run: |
            docker run --rm -v "/tmp/travis-ci-test":/tmp -w /tmp \
            ghcr.io/trufflesecurity/trufflehog:latest \
            git file:///tmp/ \
            --since-commit \
            ${{ github.event.repository.default_branch }} \
            --branch \
            'HEAD' \
            --fail \
            --no-update \
            --github-actions \
            --debug --trace \
  
```

Note: If this works, then you'd be able to use the TruffleHog action just like you showed in the "small reproduction" you provided above.

**Pull request merged**
